File: leetcode-js/98.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function (root) {
    if (!root) return false;
    let isValid = true;

    const checkBST = (node) => {
        let nodeValues = [node.val, node.val];
        if (!isValid || (!node.left && !node.right)) return nodeValues;
        let leftValues = node.left ? checkBST(node.left) : nodeValues;
        let rightValues = node.right ? checkBST(node.right) : nodeValues;
        if (!isValid) return nodeValues;
        if (node.left && (Math.max(...leftValues) >= node.val)) isValid = false;
        if (node.right && (Math.min(...rightValues) <= node.val)) isValid = false;
        if (!isValid) return nodeValues;

        return [Math.min(...leftValues, ...rightValues, node.val), Math.max(...leftValues, ...rightValues, node.val)];
    }
    checkBST(root);
    return isValid;
};

File: leetcode-go/98.go
package main

import "math"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func validateTree(root *TreeNode, isValid *bool) (int, int) {
	minLeftValue, maxLeftValue := validateTree(root.Left, isValid)
	minRightValue, maxRightValue := validateTree(root.Right, isValid)
	if !(*isValid) || maxLeftValue > root.Val || root.Val > minRightValue {
		*isValid = false
		min := int(math.Min(float64(minLeftValue), math.Min(float64(maxLeftValue), math.Min(float64(minRightValue), float64(maxRightValue)))))
		max := int(math.Max(float64(minLeftValue), math.Max(float64(maxLeftValue), math.Max(float64(minRightValue), float64(maxRightValue)))))
		return min, max
	}
	return minLeftValue, maxRightValue
}

func isValidBST(root *TreeNode) bool {
	isValid := false
	validateTree(root, &isValid)
	return isValid
}
